rosidl_generator_c/resource/idl__description.c.em

```diff
@@ -0,0 +1,140 @@
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'request_message'))
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'response_message'))
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'event_message'))
+for action in content.get_elements_of_type(Action):
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'goal'))
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'result'))
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'feedback'))
+
+  send_goal_service = extract_subinterface(type_description_msg, 'send_goal_service')
+  full_type_descriptions.append(send_goal_service)
+  full_type_descriptions.append(extract_subinterface(send_goal_service, 'request_message'))
+  full_type_descriptions.append(extract_subinterface(send_goal_service, 'response_message'))
+  full_type_descriptions.append(extract_subinterface(send_goal_service, 'event_message'))
+
+  get_result_service = extract_subinterface(type_description_msg, 'get_result_service')
+  full_type_descriptions.append(get_result_service)
+  full_type_descriptions.append(extract_subinterface(get_result_service, 'request_message'))
+  full_type_descriptions.append(extract_subinterface(get_result_service, 'response_message'))
+  full_type_descriptions.append(extract_subinterface(get_result_service, 'event_message'))
+
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'feedback_message'))
+
+def typename_to_c(typename):
+  return typename.replace('/', '__')
+
+def static_seq(varname, values):
+  """Statically define a runtime Sequence or String type."""
+  if values:
+    return f'{{{varname}, {len(values)}, {len(values)}}}'
+  return '{NULL, 0, 0}'
+}@
+
+#include "@(include_base)__struct.h"
+
+// Declare and define all type names, field names, and default values
+@[for itype_description in all_type_descriptions]@
+@{
+td_c_typename = typename_to_c(itype_description['type_name'])
+}@
+static char @(td_c_typename)__TYPE_NAME[] = "@(itype_description['type_name'])";
+@[  for field in itype_description['fields']]@
+static char @(td_c_typename)__FIELD_NAME__@(field['name'])[] = "@(field['name'])";
+@[    if field['default_value']]@
+static char @(td_c_typename)__DEFAULT_VALUE__@(field['name'])[] = "@(escape_string(field['default_value']))";
+@[    end if]@
+@[  end for]@
+
+@[end for]@
+@
+/// Define all arrays of Fields
+@[for itype_description in all_type_descriptions]@
+@{
+td_c_typename = typename_to_c(itype_description['type_name'])
+}@
+
+@[  if itype_description['fields']]@
+static rosidl_runtime_c__type_description__Field @(td_c_typename)__FIELDS[] = {
+@[    for field in itype_description['fields']]@
+  {
+    @(static_seq(f"{td_c_typename}__FIELD_NAME__{field['name']}", field['name'])),
+    {
+      @(field['type']['type_id']),  // TODO(ek) should this be named to FieldType const, or is value ok?
```

Comment:
Are we OK with number value here? We don't have the named constant at this point, but could try to do a reverse lookup in `rosidl_generator_type_description`

Reply:
I'd say that the named constant would be nice to have here, it would be easier to understand.  So I'd suggest giving the lookup a try.  If it turns out to be too complicated or otherwise not possible, we can stick with the constant.

Reply:
This is still open.  If you have some time to look into it, then I think the constant would be preferred.  If not, I'm OK with leaving it as the value, but I'll suggest:

1.  Expanding the TODO comment on what would need to be done, and
2. Putting that comment in the `empy` section, so this comment doesn't get expanded for every single message.

Reply:
Yep, done now. I hadn't got to it last night